File: ipcalc.php
# ipcalc.php
#!/usr/bin/env php
<?php

function ipToInt($ip) {
    $parts = explode('.', $ip);
    return ($parts[0] << 24) | ($parts[1] << 16) | ($parts[2] << 8) | $parts[3];
}

function intToIP($num) {
    return (($num >> 24) & 0xff) . '.' . (($num >> 16) & 0xff) . '.' . (($num >> 8) & 0xff) . '.' . ($num & 0xff);
}

function prefixToMask($prefix) {
    return (~0 << (32 - $prefix)) & 0xffffffff;
}

function maskToPrefix($mask) {
    return substr_count(decbin($mask), '1');
}

function ipBinary($num) {
    return implode('.', array_map(function($b) { return str_pad(decbin($b), 8, '0', STR_PAD_LEFT); },
        [($num >> 24) & 0xff, ($num >> 16) & 0xff, ($num >> 8) & 0xff, $num & 0xff]));
}

class IPCalc {
    public $ipInt, $maskInt, $prefix, $networkInt, $broadcast, $wildcard;
    public $totalHosts, $usableHosts, $firstUsable, $lastUsable;

    public function __construct($input, $mask = null) {
        if (strpos($input, '/') !== false) {
            list($ipPart, $prefixStr) = explode('/', $input);
            $this->ipInt = ipToInt($ipPart);
            $this->prefix = (int)$prefixStr;
            $this->maskInt = prefixToMask($this->prefix);
        } elseif ($mask !== null) {
            $this->ipInt = ipToInt($input);
            if (strpos($mask, '.') !== false) {
                $this->maskInt = ipToInt($mask);
                $this->prefix = maskToPrefix($this->maskInt);
            } else {
                $this->prefix = (int)$mask;
                $this->maskInt = prefixToMask($this->prefix);
            }
        } else {
            throw new Exception("Invalid input");
        }
        $this->networkInt = $this->ipInt & $this->maskInt;
        $this->broadcast = $this->networkInt | (~$this->maskInt & 0xffffffff);
        $this->wildcard = ~$this->maskInt & 0xffffffff;
        $this->totalHosts = 1 << (32 - $this->prefix);
        $this->usableHosts = max(0, $this->totalHosts - 2);
        if ($this->networkInt == $this->broadcast) {
            $this->firstUsable = null;
            $this->lastUsable = null;
        } else {
            $this->firstUsable = $this->networkInt + 1;
            $this->lastUsable = $this->broadcast - 1;
        }
    }

    public function toArray() {
        return [
            'ip' => intToIP($this->ipInt),
            'mask' => intToIP($this->maskInt),
            'prefix' => $this->prefix,
            'network' => intToIP($this->networkInt),
            'broadcast' => intToIP($this->broadcast),
            'wildcard' => intToIP($this->wildcard),
            'first_usable' => $this->firstUsable !== null ? intToIP($this->firstUsable) : null,
            'last_usable' => $this->lastUsable !== null ? intToIP($this->lastUsable) : null,
            'total_hosts' => $this->totalHosts,
            'usable_hosts' => $this->usableHosts,
        ];
    }

    public function subnets($newPrefix) {
        if ($newPrefix <= $this->prefix) throw new Exception("New prefix must be larger than original");
        $num = 1 << ($newPrefix - $this->prefix);
        $step = 1 << (32 - $newPrefix);
        $results = [];
        for ($i = 0; $i < $num; $i++) {
            $net = $this->networkInt + $i * $step;
            $sub = new self(intToIP($net), (string)$newPrefix);
            $results[] = [
                'network' => intToIP($sub->networkInt),
                'prefix' => $sub->prefix,
                'first_usable' => $sub->firstUsable !== null ? intToIP($sub->firstUsable) : 'none',
                'last_usable' => $sub->lastUsable !== null ? intToIP($sub->lastUsable) : 'none',
            ];
        }
        return $results;
    }
}

$opts = getopt("", ["subnet:", "bin", "json", "help"]);
if (isset($opts['help'])) {
    echo "Usage: php ipcalc.php <ip>[/prefix] [mask] [--subnet PREFIX] [--bin] [--json]\n";
    exit(0);
}
$input = $argv[1] ?? null;
$mask = $argv[2] ?? null;
if (!$input) {
    fwrite(STDERR, "Usage: php ipcalc.php <ip>[/prefix] [mask] [options]\n");
    exit(1);
}

try {
    $calc = new IPCalc($input, $mask);
} catch (Exception $e) {
    fwrite(STDERR, "Error: " . $e->getMessage() . "\n");
    exit(1);
}

$result = $calc->toArray();
if (isset($opts['bin'])) {
    $result['binary_ip'] = ipBinary($calc->ipInt);
    $result['binary_mask'] = ipBinary($calc->maskInt);
    $result['binary_network'] = ipBinary($calc->networkInt);
}
$subnets = [];
if (isset($opts['subnet'])) {
    try {
        $subnets = $calc->subnets((int)$opts['subnet']);
    } catch (Exception $e) {
        fwrite(STDERR, "Subnet error: " . $e->getMessage() . "\n");
        exit(1);
    }
}

if (isset($opts['json'])) {
    if (!empty($subnets)) $result['subnets'] = $subnets;
    echo json_encode($result, JSON_PRETTY_PRINT) . "\n";
    exit(0);
}

echo "\nIP Calculator\n";
echo "IP Address:     {$result['ip']}\n";
echo "Subnet Mask:    {$result['mask']} (/{$result['prefix']})\n";
echo "Network:        {$result['network']}\n";
echo "Broadcast:      {$result['broadcast']}\n";
echo "Wildcard Mask:  {$result['wildcard']}\n";
if ($result['first_usable']) {
    echo "First Usable:   {$result['first_usable']}\n";
    echo "Last Usable:    {$result['last_usable']}\n";
} else {
    echo "First Usable:   (none)\n";
    echo "Last Usable:    (none)\n";
}
echo "Total Hosts:    {$result['total_hosts']}\n";
echo "Usable Hosts:   {$result['usable_hosts']}\n";

if (isset($opts['bin'])) {
    echo "\nBinary:\n";
    echo "IP:      {$result['binary_ip']}\n";
    echo "Mask:    {$result['binary_mask']}\n";
    echo "Network: {$result['binary_network']}\n";
}
if (!empty($subnets)) {
    echo "\nSubnets (/" . $opts['subnet'] . "):\n";
    foreach ($subnets as $s) {
        echo "  {$s['network']}/{$s['prefix']}   ({$s['first_usable']} - {$s['last_usable']})\n";
    }
}
?>

File: ipcalc.rb
# ipcalc.rb
#!/usr/bin/env ruby
require 'optparse'
require 'json'

def ip_to_int(ip)
  parts = ip.split('.')
  (parts[0].to_i << 24) | (parts[1].to_i << 16) | (parts[2].to_i << 8) | parts[3].to_i
end

def int_to_ip(num)
  [(num >> 24) & 0xff, (num >> 16) & 0xff, (num >> 8) & 0xff, num & 0xff].join('.')
end

def prefix_to_mask(prefix)
  (~0 << (32 - prefix)) & 0xffffffff
end

def mask_to_prefix(mask)
  mask.to_s(2).count('1')
end

def ip_binary(num)
  [(num >> 24) & 0xff, (num >> 16) & 0xff, (num >> 8) & 0xff, num & 0xff]
    .map { |b| b.to_s(2).rjust(8, '0') }.join('.')
end

class IPCalc
  attr_reader :ip_int, :mask_int, :prefix, :network_int, :broadcast, :wildcard,
              :total_hosts, :usable_hosts, :first_usable, :last_usable

  def initialize(input, mask = nil)
    if input.include?('/')
      ip_part, prefix_str = input.split('/')
      @ip_int = ip_to_int(ip_part)
      @prefix = prefix_str.to_i
      @mask_int = prefix_to_mask(@prefix)
    elsif mask
      @ip_int = ip_to_int(input)
      if mask.include?('.')
        @mask_int = ip_to_int(mask)
        @prefix = mask_to_prefix(@mask_int)
      else
        @prefix = mask.to_i
        @mask_int = prefix_to_mask(@prefix)
      end
    else
      raise ArgumentError, "Invalid input"
    end
    @network_int = @ip_int & @mask_int
    @broadcast = @network_int | (~@mask_int & 0xffffffff)
    @wildcard = ~@mask_int & 0xffffffff
    @total_hosts = 1 << (32 - @prefix)
    @usable_hosts = [0, @total_hosts - 2].max
    if @network_int == @broadcast
      @first_usable = nil
      @last_usable = nil
    else
      @first_usable = @network_int + 1
      @last_usable = @broadcast - 1
    end
  end

  def to_hash
    {
      ip: int_to_ip(@ip_int),
      mask: int_to_ip(@mask_int),
      prefix: @prefix,
      network: int_to_ip(@network_int),
      broadcast: int_to_ip(@broadcast),
      wildcard: int_to_ip(@wildcard),
      first_usable: @first_usable ? int_to_ip(@first_usable) : nil,
      last_usable: @last_usable ? int_to_ip(@last_usable) : nil,
      total_hosts: @total_hosts,
      usable_hosts: @usable_hosts
    }
  end

  def subnets(new_prefix)
    raise "New prefix must be larger than original" if new_prefix <= @prefix
    num = 1 << (new_prefix - @prefix)
    step = 1 << (32 - new_prefix)
    results = []
    num.times do |i|
      net = @network_int + i * step
      sub = IPCalc.new(int_to_ip(net), new_prefix.to_s)
      results << {
        network: int_to_ip(sub.network_int),
        prefix: sub.prefix,
        first_usable: sub.first_usable ? int_to_ip(sub.first_usable) : 'none',
        last_usable: sub.last_usable ? int_to_ip(sub.last_usable) : 'none'
      }
    end
    results
  end
end

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: ipcalc.rb <ip>[/prefix] [mask] [options]"
  opts.on("--subnet PREFIX", Integer, "Create subnets of this prefix length") { |v| options[:subnet] = v }
  opts.on("--bin", "Show binary representation") { options[:bin] = true }
  opts.on("--json", "Output JSON") { options[:json] = true }
end.parse!

input = ARGV[0]
mask = ARGV[1]
unless input
  puts "Usage: ipcalc.rb <ip>[/prefix] [mask] [options]"
  exit 1
end

begin
  calc = IPCalc.new(input, mask)
rescue => e
  warn "Error: #{e.message}"
  exit 1
end

result = calc.to_hash
if options[:bin]
  result[:binary_ip] = ip_binary(calc.ip_int)
  result[:binary_mask] = ip_binary(calc.mask_int)
  result[:binary_network] = ip_binary(calc.network_int)
end

subnets = []
if options[:subnet]
  begin
    subnets = calc.subnets(options[:subnet])
  rescue => e
    warn "Subnet error: #{e.message}"
    exit 1
  end
end

if options[:json]
  result[:subnets] = subnets if subnets.any?
  puts JSON.pretty_generate(result)
  exit
end

puts "\nIP Calculator"
puts "IP Address:     #{result[:ip]}"
puts "Subnet Mask:    #{result[:mask]} (/#{result[:prefix]})"
puts "Network:        #{result[:network]}"
puts "Broadcast:      #{result[:broadcast]}"
puts "Wildcard Mask:  #{result[:wildcard]}"
if result[:first_usable]
  puts "First Usable:   #{result[:first_usable]}"
  puts "Last Usable:    #{result[:last_usable]}"
else
  puts "First Usable:   (none)"
  puts "Last Usable:    (none)"
end
puts "Total Hosts:    #{result[:total_hosts]}"
puts "Usable Hosts:   #{result[:usable_hosts]}"

if options[:bin]
  puts "\nBinary:"
  puts "IP:      #{result[:binary_ip]}"
  puts "Mask:    #{result[:binary_mask]}"
  puts "Network: #{result[:binary_network]}"
end

if subnets.any?
  puts "\nSubnets (/#{options[:subnet]}):"
  subnets.each do |s|
    puts "  #{s[:network]}/#{s[:prefix]}   (#{s[:first_usable]} - #{s[:last_usable]})"
  end
end
